Public/Src/Engine/Processes/SharedOpaqueOutputLogger.cs

```diff
@@ -0,0 +1,233 @@
+    /// 
+    /// A number of root directories may be set in which case the log will record only 
+    /// those paths that fall under one of those directories.  The typical use case is to
+    /// use all shared opaque directory outputs of a process as root directories.
+    /// </summary>
+    /// <remarks>
+    /// NOTE: not thread-safe
+    /// </remarks>
+    public sealed class SharedOpaqueOutputLogger : IDisposable
+    {
+        private readonly HashSet<AbsolutePath> m_recordedPathsCache;
+        private readonly Lazy<BuildXLWriter> m_lazyBxlWriter;
+
+        /// <summary>
+        /// Absolute path of this log file.
+        /// </summary>
+        public string LogPath { get; }
+
+        /// <summary>
+        /// Only paths under these root directories will be recorded by <see cref="RecordFileWrite(PathTable, AbsolutePath)"/>
+        /// 
+        /// When <code>null</code>, all paths are recorded.
+        /// </summary>
+        [CanBeNull]
+        public IReadOnlyList<string> RootDirectories { get; }
+
+        /// <summary>
+        /// Creates a new output logger for a given process.
+        /// 
+        /// Shared opaque directory outputs of <paramref name="process"/> are used as root directories and
+        /// <see cref="GetOutputLogFileForProcess"/> is used as log base name.
+        ///
+        /// <seealso cref="SharedOpaqueOutputLogger(string, IReadOnlyList{string})"/>
+        /// </summary>
+        public SharedOpaqueOutputLogger(PipExecutionContext context, Process process, AbsolutePath logDirectory)
+            : this(
+                  GetOutputLogFileForProcess(context.PathTable, logDirectory, process),
+                  process.DirectoryOutputs.Where(d => d.IsSharedOpaque).Select(d => d.Path.ToString(context.PathTable)).ToList())
+        {
+            Contract.Requires(process != null);
+            Contract.Requires(context != null);
+            Contract.Requires(logDirectory.IsValid);
+            Contract.Requires(Directory.Exists(logDirectory.ToString(context.PathTable)));
+        }
+
+        /// <summary>
+        /// Creates a new output logger.
+        /// 
+        /// The underlying file is created only upon first write.
+        /// </summary>
+        /// <param name="logPath">File to which to save the log.</param>
+        /// <param name="rootDirectories">Only paths under one of the root directories are recorded in <see cref="RecordFileWrite(PathTable, AbsolutePath)"/>.</param>
+        public SharedOpaqueOutputLogger(string logPath, [CanBeNull] IReadOnlyList<string> rootDirectories)
+        {
+            LogPath = logPath;
+            RootDirectories = rootDirectories;
+            m_recordedPathsCache = new HashSet<AbsolutePath>();
+
+            m_lazyBxlWriter = Lazy.Create(() =>
+            {
+                Directory.CreateDirectory(Path.GetDirectoryName(LogPath));

```

Comment:
>D [](start = 16, length = 1)

Ah, I see why you made the writer lazy. For robustness, some IO exception should be caught and reported here.

Reply:
Exactly